### PR TITLE
[12.0] Add shopinvader_partner_firstname

### PR DESCRIPTION
**Pull request by simahawk**
Integrates `partner_firstname`.

### CODE REVIEW

**Review by lmignon — COMMENTED** [review]
At first read, LGTM

**Comment by simahawk** [comment]
rebased on #469 as it contains a fix for master and to be merged soon

**Comment by simahawk** [comment]
test to fix https://travis-ci.org/shopinvader/odoo-shopinvader/jobs/607689374#L3457

**Comment by simahawk** [comment]
@lmignon good for you?

**Comment by simahawk** [comment]
hmm I don't get why tests fail only in one of the jobs https://travis-ci.org/shopinvader/odoo-shopinvader/jobs/608155216?utm_medium=notification&utm_source=github_status

**Comment by simahawk** [comment]
@lmignon @sebastienbeau can you give me your opinion regarding this? 
https://github.com/shopinvader/odoo-shopinvader/pull/475/commits/6196b937437b9efd8503a1a45da9c2c50d2748c9
To me it seems more complete and it fixes the case whereas the cache does not contain one of the values.

File: shopinvader/models/shopinvader_partner.py [hunk 1]
```diff
@@ -90,14 +90,24 @@ def _is_same_partner_value(self, partner, vals):
...
-        for key in data:
-            if key not in vals:
+        skip_keys = self._is_same_partner_value_skip_keys(partner)
+        keys_to_check = []
+        for key in vals.keys():
+            if key in skip_keys:
                 continue
+            keys_to_check.append(key)
+            if key not in partner._cache:
+                partner[key]  # load missing key
```
Comment (lmignon) [inline shopinvader/models/shopinvader_partner.py]: :champagne:

**Comment by simahawk** [comment]
last build issues fixed, we can merge as it gets green.

**Comment by simahawk** [comment]
LINT tests seem to have some warnings for many modules but not for mine. I think we can safely merge and then address warnings if needed.
I'm surprised to see a red light indeed.

File: shopinvader/models/shopinvader_partner.py [hunk 2]
```diff
@@ -90,14 +90,24 @@ def _is_same_partner_value(self, partner, vals):
...
-        for key in data:
-            if key not in vals:
+        skip_keys = self._is_same_partner_value_skip_keys(partner)
+        keys_to_check = []
+        for key in vals.keys():
+            if key in skip_keys or key not in partner:
                 continue
+            keys_to_check.append(key)
+            if key not in partner._cache:
+                partner[key]  # load missing key
```
Comment (lmignon) [inline shopinvader/models/shopinvader_partner.py]:
```suggestion
                # pylint: disable=pointless-statement
                partner[key]  # load missing key
```
@simahawk IMO pylint is failing due to this line https://travis-ci.org/shopinvader/odoo-shopinvader/jobs/611453281#L355
image

File: shopinvader/models/shopinvader_partner.py [hunk 3]
```diff
@@ -90,14 +90,24 @@ def _is_same_partner_value(self, partner, vals):
...
-        data = partner._convert_to_write(partner._cache)
-        for key in data:
-            if key not in vals:
+        skip_keys = self._is_same_partner_value_skip_keys(partner)
+        keys_to_check = []
+        for key in vals.keys():
+            if key in skip_keys or key not in partner:
                 continue
+            keys_to_check.append(key)
+            if key not in partner._cache:
```
Comment (lmignon) [inline shopinvader/models/shopinvader_partner.py]: @simahawk We don't have to check if the key is already into the cache... it's done by the orm
https://github.com/odoo/odoo/blob/12.0/odoo/fields.py#L989

**Review by lmignon — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
all :green_heart: finally :)

**Comment by lmignon** [comment]
:champagne:

**Pull request merged**